### PR TITLE
Optimise merge (~2x speedup)

### PR DESCRIPTION
**Pull request by Sothatsit**
**TL;DR:** This PR reduces time spent merging from 17.8 seconds to 8.1 seconds when running `train.py` on my laptop.

### 1. Modify array in-place

I have introduced a new method, `merge_inplace`, which modifies the ids array it is given instead of producing a new array. Additionally, I modified `merge` to copy the ids array and call `merge_inplace` on the copy, which is still quicker than the previous method (11.1 seconds spent merging when copying the array as well).

### 2. Algorithm improvements

I also made algorithm improvements so that we can skip more tokens while checking for pairs. I will try to explain why the changes work. It is probably best if you read the code before trying to understand this explanation though. Consider a pair (p0, p1) being matched against (m0, m1):
- If `p0 == p1` (e.g., "AA"), we have two cases:
  1. If `p1 != m1`, then `p0 != m1` as well. We can move forward by 2.
  2. If `p1 == m1`, then `p0 == m1` as well. We can only move forward by 1.

- If `p0 != p1` (e.g., "AB"), we have two cases:
  1. If `p1 != m1`, we can preemptively check whether `p0 != m1`, and potentially move forward by 2.
  2. If `p1 == m1` but `p0 != m0`, then `p0 != m1` (as we know p0 != p1). We can move forward by 2.

## Philosophy

I tried to keep the changes in plain Python to keep this minimal and readable. Alternatively, we could definitely optimise `merge` and `get_stats` a lot more using Numba and/or NumPy. Although, that might not fit with the spirit of this library.

## Testing

I have also added some tests for `merge` in `tests/test_merge.py`.

### CODE REVIEW

**Comment by Sothatsit** [comment]
Reading other issues on the idea for progressing this project, this probably will never be merged. Adios 🫡
